dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/InstrumentedSslConnectionSocketFactory.java

```diff
@@ -0,0 +1,206 @@
+
+    InstrumentedSslConnectionSocketFactory(
+            String clientName,
+            DialogueClientMetrics dialogueClientMetrics,
+            SSLSocketFactory socketFactory,
+            String[] supportedProtocols,
+            String[] supportedCipherSuites,
+            HostnameVerifier hostnameVerifier,
+            Supplier<Socket> rawSocketCreator) {
+        super(socketFactory, supportedProtocols, supportedCipherSuites, hostnameVerifier);
+        this.supportedProtocols = supportedProtocols;
+        this.supportedCipherSuites = supportedCipherSuites;
+        this.clientName = clientName;
+        this.rawSocketCreator = rawSocketCreator;
+        this.connectTimerSuccess = dialogueClientMetrics
+                .connectionConnect()
+                .clientName(clientName)
+                .result(ConnectionConnect_Result.SUCCESS)
+                .build();
+        this.connectTimerFailure = dialogueClientMetrics
+                .connectionConnect()
+                .clientName(clientName)
+                .result(ConnectionConnect_Result.FAILURE)
+                .build();
+    }
+
+    @Override
+    public Socket createSocket(HttpContext _context) {
+        return rawSocketCreator.get();
+    }
+
+    @Override
+    public Socket connectSocket(
+            Socket socket,
+            HttpHost host,
+            InetSocketAddress remoteAddress,
+            InetSocketAddress localAddress,
+            Timeout connectTimeout,
+            Object attachment,
+            HttpContext context)
+            throws IOException {
+        Preconditions.checkNotNull(host, "host is required");
+        Preconditions.checkNotNull(remoteAddress, "remoteAddress is required");
+        Socket sock = socket != null ? socket : createSocket(context);
+        if (localAddress != null) {
+            sock.bind(localAddress);
+        }
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "Connecting socket to {} ({}) with timeout {}",
+                        SafeArg.of("clientName", clientName),
+                        UnsafeArg.of("remoteAddress", remoteAddress),
+                        SafeArg.of("connectTimeout", connectTimeout));
+            }
+            // Run this under a doPrivileged to support lib users that run under a SecurityManager this allows granting
+            // connect permissions only to this library
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Object>) () -> {
+                    doConnect(
+                            sock,
+                            remoteAddress,
+                            Timeout.defaultsToDisabled(connectTimeout).toMillisecondsIntBound());
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof IOException) {
```

Comment:
Clarification: would we always want to throw our `SafeIoException` with the additional clientName info?

Reply:
nope -- the original exception type (as a subtype of IOException, like socketexception, etc) is meaningful